src/cargo/util/toml.rs

```diff
@@ -749,34 +754,45 @@ fn normalize(lib: &Option<TomlLibTarget>,
              examples: &[TomlExampleTarget],
              tests: &[TomlTestTarget],
              benches: &[TomlBenchTarget],
-             metadata: &Metadata) -> Vec<Target> {
+             metadata: &Metadata,
+             warnings: &mut Vec<String>) -> CargoResult<Vec<Target>> {
     fn configure(toml: &TomlTarget, target: &mut Target) {
-        let t2 = target.clone();
-        target.set_tested(toml.test.unwrap_or(t2.tested()))
-              .set_doc(toml.doc.unwrap_or(t2.documented()))
-              .set_doctest(toml.doctest.unwrap_or(t2.doctested()))
-              .set_benched(toml.bench.unwrap_or(t2.benched()))
-              .set_harness(toml.harness.unwrap_or(t2.harness()))
-              .set_for_host(toml.plugin.unwrap_or(t2.for_host()));
-    }
-
-    fn lib_target(dst: &mut Vec<Target>, l: &TomlLibTarget,
-                  metadata: &Metadata) {
-        let path = l.path.clone().unwrap_or_else(|| {
+        if let Some(test) = toml.test { target.set_tested(test); };
+        if let Some(doc) = toml.doc { target.set_doc(doc); };
+        if let Some(doctest) = toml.doctest { target.set_doctest(doctest); };
+        if let Some(bench) = toml.bench { target.set_benched(bench); };
+        if let Some(harness) = toml.harness { target.set_harness(harness); };
+        if let Some(plugin) = toml.plugin { target.set_for_host(plugin); };
+    }
+
+    fn lib_target(l: &TomlLibTarget,
+                  metadata: &Metadata,
+                  warnings: &mut Vec<String>) -> CargoResult<Target> {
+        let path = l.path.clone().unwrap_or(
             PathValue::Path(Path::new("src").join(&format!("{}.rs", l.name())))
-        });
-        let crate_types = l.crate_type.clone().and_then(|kinds| {
-            kinds.iter().map(|s| LibKind::from_str(s))
-                 .collect::<CargoResult<_>>().ok()
-        }).unwrap_or_else(|| {
-            vec![if l.plugin == Some(true) {LibKind::Dylib} else {LibKind::Lib}]
-        });
-
-        let mut target = Target::lib_target(&l.name(), crate_types.clone(),
+        );
+        let crate_types = match l.crate_type.clone() {
+            Some(kinds) =>
+                // For now, merely warn about invalid crate types.
+                // In the future, it might be nice to make them errors.
+                kinds.iter().filter_map(|s| {
+                         let kind = LibKind::from_str(s);
+                         if let Err(ref error) = kind {
+                             warnings.push(format!("warning: {}", error))
+                         }
+                         kind.ok()
+                     }).collect(),
```

Comment:
This indentation is a little off, could this follow the same pattern as the rest of Cargo?